lib/class-wp-duotone-gutenberg.php

```diff
@@ -0,0 +1,375 @@
+				// Build a unique slug for the filter based on the array of colors.
+				$slug = wp_unique_id( sanitize_key( implode( '-', $duotone_attr ) . '-' ) );
+
+				$filter_data = array(
+					'slug'   => $slug,
+					'colors' => $duotone_attr,
+				);
+				// Build a customized CSS filter property for unique slug.
+				$declaration_value = gutenberg_get_duotone_filter_property( $filter_data );
+
+				WP_Duotone_Gutenberg::$output[ $slug ] = $filter_data;
+			}
+		} elseif ( $has_global_styles_duotone ) {
+			$slug = WP_Duotone_Gutenberg::$global_styles_block_names[ $block['blockName'] ];
+
+			// Utilize existing preset CSS custom property.
+			$declaration_value = WP_Duotone_Gutenberg::get_css_var( $slug );
+
+			WP_Duotone_Gutenberg::$output[ $slug ] = WP_Duotone_Gutenberg::$global_styles_presets[ $slug ];
+		}
+
+		// - Applied as a class attribute to the block wrapper.
+		// - Used as a selector to apply the filter to the block.
+		$filter_id = gutenberg_get_duotone_filter_id( array( 'slug' => $slug ) );
+
+		// Build the CSS selectors to which the filter will be applied.
+		$selector = WP_Theme_JSON_Gutenberg::scope_selector( '.' . $filter_id, $duotone_support );
+
+		// We only want to add the selector if we have it in the output already, essentially skipping 'unset'.
+		if ( array_key_exists( $slug, WP_Duotone_Gutenberg::$output ) ) {
+			WP_Duotone_Gutenberg::$output[ $slug ]['selector'] = $selector;
+		}
+
+		// Calling gutenberg_style_engine_get_stylesheet_from_css_rules ensures that
+		// the styles are rendered in an inline for block supports because we're
+		// using the `context` option to instruct it so.
+		gutenberg_style_engine_get_stylesheet_from_css_rules(
+			array(
+				array(
+					'selector'     => $selector,
+					'declarations' => array(
+						// !important is needed because these styles
+						// render before global styles,
+						// and they should be overriding the duotone
+						// filters set by global styles.
+						'filter' => $declaration_value . ' !important',
+					),
+				),
+			),
+			array(
+				'context' => 'block-supports',
+			)
+		);
+
+		// Like the layout hook, this assumes the hook only applies to blocks with a single wrapper.
+		return preg_replace(
+			'/' . preg_quote( 'class="', '/' ) . '/',
+			'class="' . $filter_id . ' ',
+			$block_content,
+			1
+		);
+	}
+}
+
+add_action( 'wp_loaded', array( 'WP_Duotone_Gutenberg', 'set_global_styles_presets' ), 10 );
+add_action( 'wp_loaded', array( 'WP_Duotone_Gutenberg', 'set_global_style_block_names' ), 10 );
+// Remove WordPress core filter to avoid rendering duplicate support elements.
+remove_filter( 'render_block', 'wp_render_duotone_support', 10, 2 );
+add_filter( 'render_block', array( 'WP_Duotone_Gutenberg', 'render_duotone_support' ), 10, 2 );
+add_action( 'wp_enqueue_scripts', array( 'WP_Duotone_Gutenberg', 'output_global_styles' ), 11 );
+add_action( 'wp_footer', array( 'WP_Duotone_Gutenberg', 'output_footer_assets' ), 10 );
```

Comment:
I think these probably still belong in block-supports/duotone.php

Reply:
Agreed 👍

Reply:
That about sums it up for me. It's also then consistent with all the other block supports and where they manage their filters.

Reply:
> For my own learning, can you explain why? 

From my perspective, every action is a side-effect which means just importing the duotone class with these would cause the actions to be added to the page. It's expected that the block-supports files will have these side-effects, but it isn't expected to import a class and have side-effects happen.